File: src/lib/scala/Slate.Common/src/main/scala/slate/common/console/TextType.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.common.console

abstract class TextType(val color:String, val upperCase:Boolean ) {

  def format(text:String):String = {
    val checkedText = Option(text).getOrElse("")
    if ( upperCase ) checkedText.toUpperCase() else checkedText
  }
}


case object Title     extends TextType( Console.BLUE  , true  )
case object Subtitle  extends TextType( Console.CYAN  , false )
case object Url       extends TextType( Console.BLUE  , false )
case object Important extends TextType( Console.RED   , false )
case object Highlight extends TextType( Console.YELLOW, false )
case object Success   extends TextType( Console.GREEN , false )
case object Error     extends TextType( Console.RED   , false )
case object Text      extends TextType( Console.WHITE , false )

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/console/TextType.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common.console

sealed class TextType(val name:String, val color: String, private val upperCase: Boolean, val format:Boolean = true) {

    object Title     : TextType("title"    , Colors.BLUE, true)
    object Subtitle  : TextType("subtitle" , Colors.CYAN, false)
    object Url       : TextType("url"      , Colors.BLUE, false)
    object Important : TextType("important", Colors.RED, false)
    object Highlight : TextType("highlight", Colors.YELLOW, false)
    object Success   : TextType("success"  , Colors.GREEN, false)
    object Failure   : TextType("failure"  , Colors.RED, false)
    object Text      : TextType("text"     , Colors.WHITE, false)
    object Label     : TextType("label"    , Colors.WHITE, false)
    object NoFormat  : TextType("none"     , "",false)
    object NewLine   : TextType("newline"  , "",false)
    object Tab       : TextType("tab"      , "",false)
    object Raw       : TextType("raw"      , "",false, false)

    fun format(text: String): String {
        return if (upperCase) text.toUpperCase() else text
    }


    companion object {
        /**
         * Converts the string representation of a semantic text to the strongly typed object
         *
         * @param mode
         */
        fun parse(mode: String): TextType {
            return when (mode.toLowerCase()) {
                Title    .name -> Title
                Subtitle .name -> Subtitle
                Url      .name -> Url
                Important.name -> Important
                Highlight.name -> Highlight
                Success  .name -> Success
                Failure  .name -> Failure
                Text     .name -> Text
                NoFormat .name -> NoFormat
                NewLine  .name -> NewLine
                Tab      .name -> Tab
                else           -> Text
            }
        }
    }
}
